ons_ras_common/ons_asyncio.py

```diff
@@ -0,0 +1,139 @@
+#                                                                            #
+##############################################################################
+#
+#   ONSAsyncIO wraps generic endpoint access routines
+#
+##############################################################################
+import crochet
+import treq
+from json import loads, dumps
+from twisted.internet import defer
+#
+#   How long (by default) do we wait for an endpoint call before we timeout?
+#
+DEFAULT_TIMEOUT = 3
+
+
+class ONSAsyncIO(object):
+    """
+    This class is designed to take all the work out of accessing the case service. Initially it
+    should be able to validate and log events against the case service and also query the event
+    service for specific combinations of events. (for example to determine case status)
+    """
+    def __init__(self, env):
+        self._env = env
+        self._bases = None
+
+    def activate(self):
+        """
+        Set up a base URL for future access to the case service (currently via the gateway)
+        """
+        self._bases = {}
+
+    def get_base(self, endpoint):
+        """
+        Get a base reference for a given endpoint. Initially we're pointing at the gateway but allow
+        provision for holding direct routes to multiple micro-service, either by static routing or by
+        route discovery from the Gateway.
+
+        :param endpoint: endpoint we're about to call
+        :return: the base reference for the endpoint
+        """
+        if endpoint not in self._bases:
+            self._bases[endpoint] = '{}://{}:{}'.format(self._env.api_protocol, self._env.api_host, self._env.api_port)
+
+        return self._bases[endpoint]
+
+    @crochet.run_in_reactor
+    def get_route(self, endpoint):
+        """
+        Generic routine to hit a remote endpoint. This is set to run in the twisted reactor and will
+        be called with a 'crochet' wait. So it will either return the results of the call as a Python
+        structure after decoding with JSON, or it will raise an error (or timeout error).
+
+        :param endpoint: The URL we want to hit
+        :return: The decoded results of the hit
+        """
+        def check(response):
+            """
+            Make sure the call succeeded, otherwise raise an error which will abort the pipeline
+
+            :param response: A deferred response object
+            :return: A deferred response object
+            """
+            if response.code != 200:
+                raise Exception('Call to "{}" yielded error code "{}"'.format(url, response.code))
+            return response
+
+        def json(response):
+            """
+            Decode the response (from bytes to string) then convert to a Python dict
+
+            :param response: Response object
+            :return: results as a dict
+            """
+            return loads(response.decode())
+
+        def log(failure):
```

Comment:
Why are these logging helper methods needed on each class? Surely logging is a distinct aspect, unrelated to the functionality of the class? We could access loggers in the idiomatic Python way, which would obviate the need for such methods.